docs/sphinx/api/default_ops.rst

```diff
@@ -541,4 +541,123 @@ This is a non-linear transformation, and no template overloads are available.
     .. code-block:: cpp
 
         cudaq::qubit qubit;
-        my(qubit);
\ No newline at end of file
+        my(qubit);
+
+
+User-Defined Custom Operations
+==============================
+
+Users can define a custom quantum operation by its unitary matrix. First use 
+the API to register a custom operation, outside of a CUDA-Q kernel. Then the 
+operation can be used within a CUDA-Q kernel like any of the built-in operations
+defined above.
+Custom operations are supported on qubits only (`qudit` with `level = 2`).
+
+.. tab:: Python
+
+    The :code:`cudaq.register_operation` API accepts an identifier string for 
+    the custom operation and its unitary matrix. The matrix can be a `list` or
+    `numpy` array of complex numbers. A 1D matrix is interpreted as row-major.
+    
+
+    .. code-block:: python
+
+        import cudaq
+        import numpy as np
+
+        cudaq.register_operation("custom_h", 1. / np.sqrt(2.) * np.array([1, 1, 1, -1]))
+
+        cudaq.register_operation("custom_x", np.array([0, 1, 1, 0]))
+
+        @cudaq.kernel
+        def bell():
+            qubits = cudaq.qvector(2)
+            custom_h(qubits[0])
+            custom_x.ctrl(qubits[0], qubits[1])
+
+        cudaq.sample(bell).dump()
+
+        
+.. tab:: C++
+
+    The macro :code:`CUDAQ_REGISTER_OPERATION` accepts a unique name for the 
+    operation, the number of target qubits, the number of rotation parameters 
+    (can be 0), and the unitary matrix as a 1D row-major `std::vector<complex>` 
+    representation.
+    
+    .. code-block:: cpp
+
+        #include <cudaq.h>
+
+        CUDAQ_REGISTER_OPERATION(custom_h, 1, 0,
+                                {M_SQRT1_2, M_SQRT1_2, M_SQRT1_2, -M_SQRT1_2})
+
+        CUDAQ_REGISTER_OPERATION(custom_x, 1, 0, {0, 1, 1, 0})
+
+        __qpu__ void bell_pair() {
+            cudaq::qubit q, r;
+            custom_h(q);
+            custom_x<cudaq::ctrl>(q, r);
+        }
+
+        int main() {
+            auto counts = cudaq::sample(bell_pair);
+            for (auto &[bits, count] : counts) {
+                printf("%s\n", bits.data());
+            }
+        }
+
+
+For multi-qubit operations, the matrix is interpreted with MSB qubit ordering, i.e. big-endian convention.
+
+
+.. tab:: Python
+
+    .. code-block:: python
+
+        import cudaq
+        import numpy as np
+        
+        cudaq.register_operation("my_cnot", np.array([1, 0, 0, 0, 
+                                                      0, 1, 0, 0, 
+                                                      0, 0, 0, 1, 
+                                                      0, 0, 1, 0]))
```

Comment:
Maybe this should be a topic for a different PR, but why do we have to say that a custom operation like this requires controls (i.e. `my_cnot(control, target)`). It seems like a 4x4 matrix should have exactly 2 targets.

Reply:
In other words, since the 4x4 matrix was fully realized, it no longer needs to have "controls" specified in my opinion. I believe the only reason "controls" are distinctly specified from targets is that the underlying simulator can optimize the gate application by not reading or writing to half of the state vector (for 1 control). But if we are supplying a full 4x4 matrix, then I would imagine the underlying simulator has to read/write the full state vector.

So I would expect the invocation of `my_cnot` to be: `my_cnot(qubits[0:2])`.

Reply:
`my_cnot(qubits[0:2])` - this should work, I called out control and target since it is a controlled-NOT gate. May be not the best example, but I chose one from standard gate set...